File: 1.两数之和.py
#
# @lc app=leetcode.cn id=1 lang=python
#
# [1] 两数之和
#

# @lc code=start
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        # for i in range(len(nums)):
        #     for j in range(i+1, len(nums)):
        #        if nums[i]+nums[j] == target and i != j:
        #             return [i,j]
        tracker_dict = {}
        for i in range(len(nums)):
            dif = target - nums[i]
            if tracker_dict.get(dif) != None:
                return [tracker_dict.get(dif), i]
            else:
                tracker_dict[nums[i]] = i
    
# @lc code=end

File: 1.两数之和.java
import java.util.HashMap;

/*
 * @lc app=leetcode.cn id=1 lang=java
 *
 * [1] 两数之和
 */

// @lc code=start
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // int[] res = new int[2];
        // outer:
        // for(int i = 0; i < nums.length; i++){
        //     for(int j = i+1; j < nums.length; j++){
        //         if(nums[i]+nums[j]==target && i != j){
        //             res[0] = i;
        //             res[1] = j;
        //             break outer;
        //         }
        //     }
        // }
        // return res;

        int[] res = new int[2];
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        outer:
        for(int i = 0; i < nums.length; i++){
            int dif  = target - nums[i];
            if (map.get(dif) != null){
                res[0] = map.get(dif);
                res[1] = i;
                break outer;
            } else {
                map.put(nums[i], i);
            }
        }
        return res;
    }
}
// @lc code=end

